lib/rules/no-underscore-dangle.js

```diff
@@ -195,13 +205,82 @@ module.exports = {
         }
 
+        /**
+         * Check if node has dangling underscore or if node is type of ArrayPattern check its elements recursively
+         * @param {ASTNode} node node to evaluate
+         * @param {string} parentNodeType the ASTNode['type'] of the node parent
+         * @returns {void}
+         * @private
+         */
+        function deepCheckDestructured(node, parentNodeType) {
+            let identifier;
+
+            if (!node || !node.type) {
+                return;
+            }
+
+            switch (node.type) {
+                case "ArrayPattern":
+                    node.elements.forEach(element => deepCheckDestructured(element, "ArrayPattern"));
+                    break;
+                case "ObjectPattern":
+                    node.properties.forEach(property => deepCheckDestructured(property, "ObjectPattern"));
+                    break;
+                case "RestElement":
+                    deepCheckDestructured(node.argument, parentNodeType);
+                    break;
+                case "Property":
+                    deepCheckDestructured(node.value, "ObjectPattern");
+                    break;
+                case "Identifier":
+                    identifier = node.name;
+                    break;
+                default:
+                    break;
+            }
+
+            const isFromDestructuredObject = parentNodeType === "ObjectPattern" && !allowInObjectDestructuring;
+            const isFromDestructuredArray = parentNodeType === "ArrayPattern" && !allowInArrayDestructuring;
+            const hasDisallowedDestructuring = isFromDestructuredObject || isFromDestructuredArray;
+
+            if (
+                identifier &&
+                hasDisallowedDestructuring &&
+                hasDanglingUnderscore(identifier) &&
+                !isSpecialCaseIdentifierInVariableExpression(identifier) &&
+                !isAllowed(identifier)
+            ) {
+                context.report({
+                    node,
+                    messageId: "unexpectedUnderscore",
+                    data: {
+                        identifier
+                    }
+                });
+            }
+        }
+
         /**
          * Check if variable expression has a dangling underscore
          * @param {ASTNode} node node to evaluate
          * @returns {void}
          * @private
          */
         function checkForDanglingUnderscoreInVariableExpression(node) {
+            const shouldDeepCheck = node && node.type && node.type === "VariableDeclarator" && node.id.type;
+
+            if (shouldDeepCheck && node.id.type === "ArrayPattern" && Array.isArray(node.id.elements)) {
+                node.id.elements.forEach(element => {
+                    deepCheckDestructured(element, node.id.type);
+                });
+            }
+
+            if (shouldDeepCheck && node.id?.type === "ObjectPattern" && Array.isArray(node.id.properties)) {
```

Comment:
Tests on Node 12 are failing because of optional chaining syntax.

Same as the previous comment, `Array.isArray(node.id.properties)` seems unnecessary.